components/fxa-client/src/auth.rs

```diff
@@ -51,41 +42,47 @@ impl FirefoxAccount {
     ///   - `metrics` - optionally, additional metrics tracking parameters.
     ///       - These will be included as query parameters in the resulting URL.
     #[handle_error(Error)]
-    pub fn begin_oauth_flow<T: AsRef<str>>(
+    pub fn connect_with_oauth<T: AsRef<str>>(
         &self,
         // Allow both &[String] and &[&str] since UniFFI can't represent `&[&str]` yet,
         scopes: &[T],
         entrypoint: &str,
         metrics: Option<MetricsParams>,
-    ) -> ApiResult<String> {
+    ) -> ApiResult<()> {
         let scopes = scopes.iter().map(T::as_ref).collect::<Vec<_>>();
-        self.internal
-            .lock()
-            .begin_oauth_flow(&scopes, entrypoint, metrics)
+
+        // Lock self.internal to determine the oauth URL
+        let mut internal = self.internal.lock();
+        let (url, flow) = internal.begin_oauth_flow(&scopes, entrypoint, metrics)?;
+
+        // Release the lock while blocking on the OAuth handler
+        drop(internal);
+        let result = self.oauth_handler.perform_flow(url)?;
```

Comment:
I'm both excited and nervous to see this, and I'm really curious what the implementation on the foreign language would look like.

It would help us if we have some tests here as well, maybe simple unit tests here that mock the handler and more importantly some Android and iOS tests that try as close as possible to mimic what the consumer might try to do:
- A simple case were the consumer returns right away in the same thread
- A more complex case where a consumer blocks the thread, then resumes it later
- A case covering an error
- A case covering cancellation

Reply:
I'm actually very glad you're bringing this up.  I want to open another PR to address locking using this example, but I was worried it was a bit too obscure.  I'll try to open that one up tomorrow, let's discuss more.

Reply:
Not clear we actually want "locking" here - indeed, not clear what semantics we want in general (ie, I think I know what we don't want, just not sure what we do :) - looking forward to chatting about it though.